enterprise_gateway/services/kernels/remotemanager.py

```diff
@@ -189,6 +192,57 @@ async def start_kernel(self, *args, **kwargs):
         self.parent.kernel_session_manager.create_session(kernel_id, **kwargs)
         return kernel_id
 
+    async def restart_kernel(self, kernel_id):
+        kernel = self.get_kernel(kernel_id)
+        self.log.debug(f"Current value of the kernel.restarting: {kernel.restarting}")
+        if kernel.restarting:  # assuming duplicate request.
+            await self.wait_for_restart_finish(kernel_id, "restart")
+            self.log.info(
+                f"Done with waiting for restart to complete. Current value of kernel.restarting: {kernel.restarting}. "
+                f"Skipping kernel restart."
+            )
+            return
+        self.log.info("Going ahead to process kernel restart request.")
+        try:
+            kernel.restarting = True  # Moved in out of RemoteKernelManager
+            await super().restart_kernel(kernel_id)
+        finally:
+            self.log.debug("Resetting kernel.restarting flag to False.")
+            kernel.restarting = False
+
+    async def shutdown_kernel(self, kernel_id, now=False, restart=False):
+        kernel = self.get_kernel(kernel_id)
+        self.log.debug(f"Current value of the Kernel Restarting: {kernel.restarting}")
+        if kernel.restarting:
+            await self.wait_for_restart_finish(kernel_id, "shutdown")
+        try:
+            await super().shutdown_kernel(kernel_id, now, restart)
+        except KeyError as ke:  # this is hit for multiple shutdown request.
+            self.log.exception(f"Exception while shutting Kernel: {kernel_id}: {ke}")
```

Comment:
```suggestion
            self.log.exception(f"Exception while shutting down kernel: '{kernel_id}': {ke}")
```
Shouldn't this be re-raised?  Or perhaps only if 'restarting' == False?

Reply:
this usually happens when we have sent duplicate kernel shutdown request while the kernel was still restarting.
 I will testing this by raising the exception and get back with the behaviour.

Reply:
Hi @kevin-bates : I tested various scenarios for this: 
1. The default behaviour of the JEG when multiple `shutdown` request is sent for the same kernel ID is except the first requests which returns 204 after shutting down the kernel, the other requests get en exception while `popping` out the `kernel_id ` from `MultiKernelManager._kernel` dict. This leads to `KeyError: 'b0b149e8-8a22-48db-b8b6-ed90e441ac41'` and the final response thrown to the clients is HTTP 500: `500 DELETE /api/kernels/b0b149e8-8a22-48db-b8b6-ed90e441ac41`.

2. With the current code change in place, we are handling the exception gracefully and returning a HTTP response 204. This behaviour is different from the point 1 above but this is similar to to the scenario where if you had sent a `shutdown` request for the `kernel_id` which does not exist. https://github.com/jupyter-server/jupyter_server/blob/main/jupyter_server/services/kernels/kernelmanager.py#L654  

let me know how to proceed forward with this. 
I feel in order to keep the behaviour same, we can skip the graceful handling and let it throw `500` to user as before.

Reply:
Could we please update the text of the log statement to the suggested value? It doesn't adhere to the conventions of the others (lower-case `kernel`, quoted `kernel_id`, and adds 'down' to complete the action).

Reply:
Sorry, I just realized I didn't answer your question above.  

Hmm.  The timing issue you describe goes beyond the current checks made in the [`_check_kernel_id()` override](https://github.com/jupyter-server/jupyter_server/blob/main/jupyter_server/services/kernels/kernelmanager.py#L493-L497) in which we'll return a `404` if the kernel_id is unknown to the `MultiKernelManager` - correct?  (I had [fixed one such race condition a while okay related to the `_kernel_connections` list](https://github.com/jupyter/notebook/pull/5875).)

So I guess a similar issue still exists wrt to `_kernels` and the `try/except KeyError` block is a catch for that.  (You're using the latest server code with `AsyncKernelManagement` - correct?)

I like the fact that we'd no longer raise a 500 in this case, and most cases of this nature will result in a `404` due to the `_check_kernel_id()` override.  I think it could be misleading to return `204` when such requests did not delete the resource (kernel) and believe the best status, in this case, would be  `404` - since that's the truth and essentially exhibits similar behavior to when the kernel-id is not managed (which is also true).

Reply:
yes @kevin-bates ..your observations are correct and I am testing with `AsyncKernelManagement`. 
The next step here is to raise a `404` similar to [check_kernel_id](https://github.com/jupyter-server/jupyter_server/blob/main/jupyter_server/services/kernels/kernelmanager.py#L493-L497).

```
raise web.HTTPError(404, "Kernel does not exist: %s" % kernel_id)
```

Also, should I further raise a CR in `AsyncMappingKernelManager::shutdown_kernel` to handle the  `KeyError` exception in jupyter server repo?

Reply:
Under normal circumstances, I'd say, yeah, if you can reproduce it in that environment.  But _that environment_ (outside of EG) will be using the pending kernel support - in which case I'd be surprised this can be reproduced. So, it comes down to, "is it worth a patch to jupyter_client 6.x?" and I'd say, let the status quo flow and live with the fact that `500` will be returned in this relatively rare scenario.  If we find issues here when converting to provisions, we can tackle this then, but this particular area of the stack is still evolving (e.g., _pending kernels_ will likely be subsumed by the _state machine_ work).